elki-outlier/src/main/java/elki/outlier/isolation/IsolationForest.java

```diff
@@ -0,0 +1,366 @@
+ * 
+ * @author Fei Tony Liu et al. (Original Code)
+ * @author Braulio V.S. Vinces (ELKIfication)
+ *
+ * @param <V> Vector type
+ */
+@Title("IsolationForest: Isolation-Based Anomaly Detection")
+@Description("This algorithm detects anomalies purely based on the concept of isolation without employing any distance or density measure.")
+@Reference(authors = "Fei Tony Liu, Kai Ming Ting, Zhi-Hua Zhou", //
+    title = "Isolation-Based Anomaly Detection", //
+    booktitle = "ACM Transanctions on Knowledge Discovery from Data (TKDD'12)", //
+    url = "https://doi.org/10.1145/2133360.2133363", //
+    bibkey = "DBLP:journals/tkdd/LiuTZ12")
+@Alias({ "de.lmu.ifi.dbs.elki.algorithm.outlier.IsolationForest", "iforest" })
+public class IsolationForest implements OutlierAlgorithm {
+
+  /**
+   * Class logger
+   */
+  private static final Logging LOG = Logging.getLogger(IsolationForest.class);
+
+  /*
+   * The set of trees
+   */
+  protected List<Tree> trees = null;
+
+  /*
+   * The number of trees
+   */
+  protected int numTrees = 100;
+
+  /*
+   * The sub sample size
+   */
+  protected int subsampleSize = 256;
+
+  /**
+   * Random generator
+   */
+  private RandomFactory rnd;
+
+  /**
+   * Constructor.
+   *
+   * @param numTrees
+   * @param subSamplingSize
+   * @param rnd
+   */
+  public IsolationForest(int numTrees, int subSamplingSize, RandomFactory rnd) {
+    this.numTrees = numTrees;
+    this.subsampleSize = subSamplingSize;
+    this.rnd = rnd;
+  }
+
+  public OutlierResult run(Database db, Relation<? extends NumberVector> relation) {
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("IsolationForest progress", relation.size(), LOG) : null;
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+
+    // Reduce sub sample size if data is too small
+    if(relation.size() < subsampleSize) {
+      subsampleSize = relation.size();
+    }
+
+    {
+      final Random random = rnd.getSingleThreadedRandom();
+      // Generate trees
+      trees = new ArrayList<>(numTrees);
+      for(int i = 0; i < numTrees; i++) {
+        ModifiableDBIDs sample = DBIDUtil.randomSample(relation.getDBIDs(), subsampleSize, rnd);
+        trees.add(new Tree( //
+            sample, //
+            random, //
+            0, //
+            (int) FastMath.ceil(FastMath.log(relation.size()) / FastMath.log(2)), //
+            relation));
+        LOG.incrementProcessed(prog);
+      }
+      LOG.ensureCompleted(prog);
+
+      // Iterate over all objects
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance()) {
+        final double score = distributionForInstance(relation.get(iter));
+        scores.putDouble(iter, score);
+        minmax.put(score);
+        LOG.incrementProcessed(prog);
```

Comment:
Do not re-use a progress that you already "completed" above. This is a second phase, and one step in this progress is much less effort than one tree construction above.